### PR TITLE
Mpay

### PR DESCRIPTION
**Pull request by michael1011**

### CODE REVIEW

File: tools/plugins/mpay/data/route_stats.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+
+class RouteStatsFetcher:
+    _db: Database
+
+    def __init__(self, db: Database) -> None:
+        self._db = db
+
+    def get_destinations(self) -> list[str]:
+        with Session(self._db.engine) as s:
```
Comment (jackstar12) [inline tools/plugins/mpay/data/route_stats.py:46]: external session

**Review by dni — APPROVED** [review]
utACK, good pr :)

File: tools/plugins/mpay/data/router.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            except BaseException as e:
+                # TODO: penalize in db?
+                self._pl.log(f"Disregarding route {stat_route} because: {format_error(e)}")
+
+    def _transform_stat_route(
+        self, stats: RouteStats, amount: Millisatoshi, exclude_list: ExcludesPayment
+    ) -> Route:
+        for hop in stats.route:
+            if hop in exclude_list:
+                raise InExcludeListError(hop)
```
[the remark this answers is not in the document]
Comment (michael1011) [reply tools/plugins/mpay/data/router.py:61]: That list can change and have more entries after that `get_routes` call finished. But true. What we can, we should exclude there

File: tools/plugins/mpay/pay/mpay.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        dec = self._pl.rpc.decodepay(bolt11)
+
+        amount = dec["amount_msat"]
+        payment_hash = dec["payment_hash"]
+
+        already_paid = self._check_for_paid(payment_hash)
+        if already_paid is not None:
+            return already_paid
+
+        with Session(self._db.engine) as session:
```
[the remark this answers is not in the document]
Comment (michael1011) [reply tools/plugins/mpay/pay/mpay.py:48]: Here, I would leave the session internal. I really don't want it to be reused for other operations outside of this function. A `commit` or `rollback` called from the outside could leave the database in an inconsistent state and lead to corrupt routes being written to the database

**Review by jackstar12 — APPROVED** [review]

**Pull request merged**